app/src/main/java/org/mozilla/reference/browser/tabs/SyncedTabsFragment.kt

```diff
@@ -0,0 +1,271 @@
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View {
+        return inflater.inflate(R.layout.fragment_synced_tabs, container, false)
+    }
+
+    override fun onViewCreated(rootView: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(rootView, savedInstanceState)
+
+        accountManager = requireContext().components.backgroundServices.accountManager
+        syncedTabsFeature = requireContext().components.backgroundServices.syncedTabs
+
+        recyclerView = rootView.findViewById(R.id.synced_tabs_list)
+        recyclerView.layoutManager = LinearLayoutManager(requireContext())
+
+        // Listen to Sync/FxA events so we can update the UI accordingly.
+        accountManager.registerForSyncEvents(
+            SyncedTabsSyncObserver(this),
+            owner = this,
+            autoPause = true
+        )
+        accountManager.register(SyncedTabsAccountObserver(this))
+        updateUI()
+    }
+
+    override fun onResume() {
+        super.onResume()
+        // Kick off a sync in the background so we get the freshest tabs.
+        scope.launch {
+            accountManager.syncNowAsync(SyncReason.User).await()
+        }
+    }
+
+    fun updateUI() {
+        val accountManager = requireContext().components.backgroundServices.accountManager
+        val account = accountManager.authenticatedAccount()
+
+        if (account != null && !accountManager.accountNeedsReauth()) {
+            // Signed-in, no problems.
+            if (SyncEnginesStorage(requireContext()).getStatus()[SyncEngine.Tabs] == true) {
+                // Tabs engine enabled.
+                showTabs()
+            } else {
+                // Tab engine disabled
+                showInfoMessage("Please enable tab syncing.")
+            }
+        } else if (account != null && accountManager.accountNeedsReauth()) {
+            // Signed-in, need to re-authenticate.
+            showInfoMessage("Please re-authenticate.")
+        } else {
+            // Signed-out.
+            showInfoMessage("Connect with a Firefox Account.")
+        }
+    }
+
+    private fun showInfoMessage(msg: String) {
+        synced_tabs_status.text = msg
+        synced_tabs_list.visibility = View.GONE
+        synced_tabs_status.visibility = View.VISIBLE
+    }
+
+    private fun showTabs() {
+        synced_tabs_list.visibility = View.VISIBLE
+        synced_tabs_status.visibility = View.GONE
+        scope.launch {
+            val syncedTabs = syncedTabsFeature.getSyncedTabs()
+            scope.launch(Dispatchers.Main) {
+                if (syncedTabs.isEmpty() &&
+                    accountManager.authenticatedAccount()?.deviceConstellation()?.state()?.otherDevices?.isEmpty()
+                    == true) {
+                    showInfoMessage("Connect another device.")
+                } else {
+                    val adapter = SyncedTabsAdapter(
```

Comment:
When we invoke `showTabs`, we're creating a new adapter every time. This can be expensive since we may be creating a new view each time, instead of... _drum roll_ 🥁 ...recycling them! 😄 

You can hold an instance of the adapter in the fragment, and notify it when we update data.

The [`ListAdapter`](https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter) is an abstraction on top of the default RecyclerViewAdapter that might make it easier to use.

I recently wrote one [in Fenix](https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/migration/MigrationProgressActivity.kt#L102-L143) if you're looking for a... 🥁  ..reference.